lib/bolt/transport/sudoable.rb

```diff
@@ -0,0 +1,163 @@
+        input_method = implementation['input_method']
+        extra_files = implementation['files']
+
+        with_connection(target, options.fetch('_load_config', true)) do |conn|
+          conn.running_as(options['_run_as']) do
+            stdin, output = nil
+            command = []
+            execute_options = {}
+            execute_options[:interpreter] = select_interpreter(executable, target.options['interpreters'])
+            interpreter_debug = if execute_options[:interpreter]
+                                  " using '#{execute_options[:interpreter]}' interpreter"
+                                end
+            # log the arguments with sensitive data redacted, do NOT log unwrapped_arguments
+            logger.debug("Running '#{executable}' with #{arguments}#{interpreter_debug}")
+            # unpack any Sensitive data
+            arguments = unwrap_sensitive_args(arguments)
+
+            conn.with_tempdir do |dir|
+              if extra_files.empty?
+                task_dir = dir
+              else
+                # TODO: optimize upload of directories
+                arguments['_installdir'] = dir.to_s
+                task_dir = File.join(dir.to_s, task.tasks_dir)
+                dir.mkdirs([task.tasks_dir] + extra_files.map { |file| File.dirname(file['name']) })
+                extra_files.each do |file|
+                  conn.copy_file(file['path'], File.join(dir.to_s, file['name']))
+                end
+              end
+
+              remote_task_path = conn.write_executable(task_dir, executable)
+
+              if STDIN_METHODS.include?(input_method)
+                stdin = JSON.dump(arguments)
+              end
+
+              if ENVIRONMENT_METHODS.include?(input_method)
+                execute_options[:environment] = envify_params(arguments)
+              end
+
+              if conn.run_as && stdin
+                # Inject interpreter in to wrapper script and remove from execute options
+                wrapper = make_wrapper_stringio(remote_task_path, stdin, execute_options[:interpreter])
+                execute_options.delete(:interpreter)
+                remote_wrapper_path = conn.write_executable(dir, wrapper, 'wrapper.sh')
+                command << remote_wrapper_path
+              else
+                command << remote_task_path
+                execute_options[:stdin] = stdin
+              end
+              dir.chown(conn.run_as)
+
+              execute_options[:sudoable] = true if conn.run_as
+              output = conn.execute(command, execute_options)
+            end
+            Bolt::Result.for_task(target, output.stdout.string,
+                                  output.stderr.string,
+                                  output.exit_code,
+                                  task.name)
+          end
+        end
+      end
+
+      def make_wrapper_stringio(task_path, stdin, interpreter = nil)
+        if interpreter
+          StringIO.new(<<-SCRIPT)
```

Comment:
Since we've switched to requiring ruby 2.3  I think we can rely on the tilde heredoc that is indentable. 

https://infinum.co/the-capsized-eight/multiline-strings-ruby-2-3-0-the-squiggly-heredoc